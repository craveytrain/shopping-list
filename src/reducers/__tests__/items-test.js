import reducer from 'reducers/items';
import {
  addItem,
  removeItem,
  toggleItem,
  nameItem,
  categorizeItem
} from 'actions/items';
import deepFreeze from 'deep-freeze';

describe('items reducer', () => {
  describe('addItem', () => {
    it('should add an item to the list', () => {
      const nextState = reducer(
        {},
        addItem({
          name: 'Leche'
        })
      );

      expect(nextState).toMatchSnapshot();
    });

    it('should respect the id if given', () => {
      const nextState = reducer(
        {},
        addItem({
          name: 'Leche',
          id: 'leche'
        })
      );

      expect(nextState).toMatchSnapshot();
    });

    it('should add an item by string', () => {
      const nextState = reducer(
        {},
        addItem('Leche')
      );

      expect(nextState).toMatchSnapshot();
    });
  });

  describe('removeItem', () => {
    it('should remove an item from the list', () => {
      const initialState = {
        byId: {
          milk: {
            name: 'Milk'
          },
          leche: {
            name: 'Leche'
          }
        },
        allIds: [ 'milk', 'leche' ]
      };

      deepFreeze(initialState);

      const nextState = reducer(initialState, removeItem('milk'));

      expect(nextState).toMatchSnapshot();
    });
  });

  describe('toggleItem', () => {
    it('should check the selected item from the list', () => {
      const initialState = {
        byId: {
          milk: {
            name: 'Milk',
            checked: false
          },
          leche: {
            name: 'Leche',
            checked: false
          }
        },
        allIds: [ 'milk', 'leche' ]
      };

      deepFreeze(initialState);

      const nextState = reducer(initialState, toggleItem('milk'));

      expect(nextState).toMatchSnapshot();
    });

    it('should uncheck the selected item from the list', () => {
      const initialState = {
        byId: {
          milk: {
            name: 'Milk',
            checked: false
          },
          leche: {
            name: 'Leche',
            checked: true
          }
        },
        allIds: [ 'milk', 'leche' ]
      };

      deepFreeze(initialState);

      const nextState = reducer(initialState, toggleItem('leche'));

      expect(nextState).toMatchSnapshot();
    });
  });

  describe('nameItem', () => {
    it('should set the name of the selected item from the list', () => {
      const initialState = {
        byId: {
          milk: {
            name: 'Milk'
          },
          leche: {
            name: 'Leche'
          }
        },
        allIds: [ 'milk', 'leche' ]
      };

      deepFreeze(initialState);

      const nextState = reducer(initialState, nameItem('milk', 'Cow Milk'));

      expect(nextState).toMatchSnapshot();
    });
  });

  describe('categorizeItem', () => {
    it('should set the categoryId of the selected item from the list', () => {
      const initialState = {
        byId: {
          leche: {
            name: 'Leche'
          }
        },
        allIds: [ 'leche' ]
      };

      deepFreeze(initialState);

      const nextState = reducer(
        initialState,
        categorizeItem('leche', 'asdasdfadf')
      );

      expect(nextState).toMatchSnapshot();
    });
  });

  it('can be used with reduce', () => {
    const actionQueue = [
      addItem({
        id: 'milk',
        name: 'Milk'
      }),
      addItem({
        id: 'goatMilk',
        name: 'Goat Milk'
      }),
      addItem({
        name: 'Leche'
      }),
      nameItem('milk', 'Cow Milk'),
      categorizeItem('milk', 'dairy'),
      removeItem('goatMilk'),
      toggleItem('milk')
    ];

    const finalState = actionQueue.reduce(reducer, {});

    expect(finalState).toMatchSnapshot();
  });
});
