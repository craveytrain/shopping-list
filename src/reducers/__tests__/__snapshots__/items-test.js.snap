// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`items reducer addItem should add an item to the list 1`] = `
Array [
  Object {
    "id": "milk",
    "name": "Milk",
  },
  Object {
    "checked": false,
    "id": "107a8b69eed98af82d87dabc319af32faad6326f",
    "name": "Leche",
  },
]
`;

exports[`items reducer addItem should respect the id if given 1`] = `
Array [
  Object {
    "id": "milk",
    "name": "Milk",
  },
  Object {
    "checked": false,
    "id": "leche",
    "name": "Leche",
  },
]
`;

exports[`items reducer can be used with reduce 1`] = `
Array [
  Object {
    "categoryId": "dairy",
    "checked": true,
    "id": "milk",
    "name": "Cow Milk",
  },
  Object {
    "checked": false,
    "id": "107a8b69eed98af82d87dabc319af32faad6326f",
    "name": "Leche",
  },
]
`;

exports[`items reducer categorizeItem should set the categoryId of the selected item from the list 1`] = `
Array [
  Object {
    "categoryId": "asdasdfadf",
    "id": "leche",
    "name": "Leche",
  },
]
`;

exports[`items reducer nameItem should set the name of the selected item from the list 1`] = `
Array [
  Object {
    "id": "milk",
    "name": "Cow Milk",
  },
  Object {
    "id": "leche",
    "name": "Leche",
  },
]
`;

exports[`items reducer removeItem should remove an item from the list 1`] = `
Array [
  Object {
    "id": "leche",
    "name": "Leche",
  },
]
`;

exports[`items reducer setItems should set items 1`] = `
Array [
  Object {
    "name": "Leche",
  },
]
`;

exports[`items reducer toggleItem should check the selected item from the list 1`] = `
Array [
  Object {
    "checked": true,
    "id": "milk",
    "name": "Milk",
  },
  Object {
    "checked": false,
    "id": "leche",
    "name": "Leche",
  },
]
`;

exports[`items reducer toggleItem should uncheck the selected item from the list 1`] = `
Array [
  Object {
    "checked": false,
    "id": "milk",
    "name": "Milk",
  },
  Object {
    "checked": false,
    "id": "leche",
    "name": "Leche",
  },
]
`;
